reuse levelDB in-memory batch and fix batchSize reset

cendhu opened the pull request:
#### Type of change

- Bug fix
- Improvement (improvement to code, performance, etc)

#### Description

This PR

(1) Enables the reuse of leveldb in-memory batch -- reduces the memory utilization and also the burden on GC.
(2) Fixes a missing logic of batchsize reset in confighistory store `ImportXXX()`

ghaskins commented:
@christo4ferris Of relevance to you is "make linter" will now ensure any go-tool that you add to $(GOTOOLS) is present before running.

@corecode Of relevance to you is that the baseimage build now has much smarter prerequisites like you asked for.

ghaskins commented:
@bcbrock Of relevance to you: tcl is now installed in the travis environment.

@binhn This has the common definition of the baseimage release ("0.0.9") that you asked for a while back.

ghaskins commented:
@rameshthoomu

bcbrock commented:
@ghaskins Would it make sense to consolidate all testing, linting etc. done by Travis into a single `make ci-test` target? That might make it easier to maintain, and easier for developers to run the suite before making a PR.

ghaskins commented:
@bcbrock I _think_ I may have captured that in "make checks", or at least that was my intention unless I am misunderstanding your point.  However, travis.yml still calls unit/behave/linter as discrete steps because I was trying to retain @rameshthoomu's log granularity.  I will defer to @rameshthoomu on whether he would like to retain the granularity of prefer to close the loop on that particular maintenance burden.

ghaskins commented:
Fixes #1539

bcbrock commented:
@ghaskins I didn't realize that `make checks` was there; Would have assumed Travis would be using it if it were.

ghaskins commented:
My rationale for leaving it discrete steps was that @rameshthoomu went through the trouble of making the travis log collapsable per section...e.g. each step in the process end up with its own log that can be optionally hidden/shown by the user in the UI.  If I converted travis to use "make checks", the entire process would run under one log session and partially defeat his setup.  I didn't want to be the one to propose to break that UI element, even though doing so violates DRY (make checks and travis have to be manually synchronized).  I don't have a strong opinion, as I understand both sides of the argument.

ghaskins commented:
@bcbrock related to your point, I would like to see us eventually advocate that "make checks" should be part of the PR checklist.  I see it as ultimately replacing the one we have now that explicitly lists goimports, golint, etc.  However, until we implement "make linter", it's too early and should remain as it is now.

ghaskins commented on Makefile:
```diff
@@ -85,20 +108,37 @@ linter:
+	$(eval TOOL = ${subst gotool.,,${@}})
+	go get ${go.fqp.${TOOL}}
+
+$(GOPATH)/bin/%:
```
I think I can do this part better.  Hold off on merging for now.

ghaskins replied:
Nevermind, my experiment failed to produce a better approach.  Good to go as is.

srderson commented:
Would it make sense to move /baseimage_release to /images/base/release? Guess I don't fully understand why it's a top level file.

ghaskins commented:
@srderson this is a great suggestion, I admit it just simply didn't occur to me.  Will make the change and push an update.

srderson commented:
LGTM. Tested and confirmed this is working. Thanks.

cendhu committed:
reuse leveldb batch and fix batchsize reset

This commit
(1) Enables the reuse of leveldb batch -- reduces the memory utilisation
    and also the burden on GC.
(2) Fixes a missing logic of batchsize reset in confighistory store
    `ImportXXX()`

Signed-off-by: senthil <cendhu@gmail.com>
